crypto/stack/stack.c

```diff
@@ -21,10 +28,43 @@ struct stack_st {
     OPENSSL_sk_compfunc comp;
 };
 
-#undef MIN_NODES
-#define MIN_NODES       4
+/*
+ * Calculate the array growth based on the target size.
+ *
+ * The growth faction is a rational number and is defined by a numerator
+ * and a denominator.  According to Andrew Koenig in his paper "Why Are
+ * Vectors Efficient?" from JOOP 11(5) 1998, this factor should be less
+ * than the golden ratio (1.618...).
+ *
+ * We use 3/2 = 1.5 for simplicty of calculation and overflow checking.
+ * Another option 8/5 = 1.6 allows for slightly faster growth, although safe
+ * computation is more difficult.
+ *
+ * The limit to avoid overflow is spot on.  2^n-1 is congruent to 0 mod 3
+ * when n is even and 1 mod 3 when n is odd.  When n is odd, int_max = 2^n-1
+ * produces limit = 2*floor(int_max/3) which is thus low by one, hence the
+ * correction term that adds one back for this case.
+ */
+static ossl_inline size_t compute_growth(size_t target, size_t current)
+{
+    const size_t limit = (INT_MAX / 3) * 2 + INT_MAX % 3;
+    size_t sz, n = current;
+
+    while (n <= target) {
```

Comment:
BTW, why `<=` and not `<`? Whole idea behind reservation is that you're telling that you're pretty sure that `target` is all you'll ever need. Then why would you insist on 50% overhang?

Reply:
With this in mind one can actually wonder if reservation should even go through this loop. I mean if you're pretty sure that `target` is all you'll *ever* need, then why have *any* overhang? Even if it turns out that it needs extension after all [as result of single insertion], one can increment from the point of initial reservation.

Reply:
I'll see if I can figure something out so the space isn't wasted.